bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java

```diff
@@ -152,18 +161,66 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-            if (CoreItemFactory.DIMMER.equals(channelConfig.itemType)) {
+            if (isColorChannelCommand) {
+                command = OnOffType.OFF == onOffCommand ? PercentType.ZERO : PercentType.HUNDRED;
+            } else if (CoreItemFactory.DIMMER.equals(channelConfig.itemType)) {
                 zwaveCommand.value = OnOffType.ON == onOffCommand ? 255 : 0;
             } else {
                 zwaveCommand.value = onOffCommand == (channelConfig.inverted ? OnOffType.OFF : OnOffType.ON);
             }
-        } else if (command instanceof HSBType hsbTypeCommand) {
-            int[] rgb = ColorUtil.hsbToRgb(hsbTypeCommand);
-            zwaveCommand.value = String.format("%02X%02X%02X", rgb[0], rgb[1], rgb[2]);
+        }
+        // note: HSBType is a child of PercentType so we must explicitly NOT execute this block in such case
+        if (PercentType.class.equals(command.getClass()) && (command instanceof PercentType percentTypeCommand)) {
+            if (isColorChannelCommand) {
+                command = new HSBType(cachedColor.getHue(), cachedColor.getSaturation(), percentTypeCommand);
+            } else {
+                int newValue = percentTypeCommand.intValue();
+                if (channelConfig.inverted) {
+                    newValue = 100 - newValue;
+                }
+                if (CoreItemFactory.DIMMER.equals(channelConfig.itemType) && newValue == 100) {
+                    newValue = 99;
+                }
+                zwaveCommand.value = newValue;
+            }
+        }
+        if (command instanceof HSBType hsbTypeCommand) {
+            cachedColor = hsbTypeCommand;
+            HSBType hsbFull = new HSBType(hsbTypeCommand.getHue(), hsbTypeCommand.getSaturation(), PercentType.HUNDRED);
+            int[] rgb = ColorUtil.hsbToRgb(hsbFull);
+            if (channelUID.getId().contains(HEX)) { // TODO it is not obvious if there is a better way to do this
+                zwaveCommand.value = "%02X%02X%02X".formatted(rgb[0], rgb[1], rgb[2]);
+            } else {
+                Map<String, Integer> colorMap = new HashMap<>();
+                colorMap.put(RED, rgb[0]);
+                colorMap.put(GREEN, rgb[1]);
+                colorMap.put(BLUE, rgb[2]);
+                if (supportsColdWhite) {
+                    colorMap.put(COLD_WHITE, 0);
+                }
+                if (supportsWarmWhite) {
+                    colorMap.put(WARM_WHITE, 0);
+                }
+                zwaveCommand.value = colorMap;
+            }
+            if (isColorChannelCommand) {
+                // schedule brightness command(s) for dimmer channel(s)
+                PercentType brightness = hsbTypeCommand.getBrightness();
+                thing.getChannels().stream().filter(c -> CoreItemFactory.DIMMER.equals(c.getAcceptedItemType()))
+                        .forEach(c -> scheduler.submit(() -> handleCommand(c.getUID(), brightness)));
```

Comment:
Have two question sabou this:
1. A node can have multiple dimmer channels where not all of them are brightness
2. A command can lead up to two commands (HS + B) causing some possible flickering of a lightsource. Have you looked into the API to combine commands to the zwave server?

Reply:
Double checked, combining into 1 message or sending a batch of messages is not supported.